File: examples/c++/ff6dof-kine-control.cpp
/*
 * Test the control of the robot
 *
 * To execute:
 *      ./ff6dofKineControl backend - where backend can be pin (by default) or rbdl 
 * 
 */

#include <oscr/model/robot-model-pin.hpp>
#include <oscr/model/robot-model-rbdl.hpp>
#include <oscr/ik/kine-task-pose.hpp>
#include <oscr/ik/osik-solvers.hpp>
#include <oscr/tools/vector-logger.hpp>


int main(int argc, char *argv[])
{
  // Use a floating base?
  bool fbase = true;
  // Robot model
  std::string modelname = "../models/ff6dof.urdf";
  // Dummy initialization
  oscr::RobotModel *robot = new oscr::RobotModelPin();
  
  if (argc<2)
  {
    std::cout << "-- Using Pinocchio as backend by default" << std::endl;
    robot = new oscr::RobotModelPin(modelname, fbase);
  }
  else
  {
    if (!strcmp(argv[1],"pin"))
    {
      std::cout << "-- Using Pinocchio as backend" << std::endl;
      robot = new oscr::RobotModelPin(modelname, fbase);
    }
    else if (!strcmp(argv[1], "rbdl"))
    {
      std::cout << "-- Using RBDL as backend" << std::endl;
      robot = new oscr::RobotModelRbdl(modelname, fbase);
    }
    else
    {
      std::cout << "Only pin and rbdl are accepted as backends ..."
                << "ending the program ..." << std::endl;
      return -1;
    }
  }
    
  std::map<std::string, unsigned int> linkID;
  linkID = robot->mapLinkNamesIDs();

  // Joint configuration with all values to zero.
  Eigen::VectorXd q(robot->ndof()); q.setZero();
  // If there is a floating base, then set a valid null quaternion
  if (robot->hasFloatingBase())
    q[3]=1.0;
  // Set the current joint configuration
  robot->updateJointConfig(q);
  // Get the current joint configuration
  q = robot->getJointConfig();

  // Definition of a task for the arm (keep) and a task for the leg
  oscr::KineTask *taskarm = new oscr::KineTaskPose(robot, linkID["larm3"], "position");
  oscr::KineTask *taskleg = new oscr::KineTaskPose(robot, linkID["lleg3"], "position");
  taskarm->setGain(1.0);
  taskleg->setGain(1.0);

  // Desired value for the arm (keep it)
  Eigen::VectorXd xarm;
  taskarm->getSensedValue(xarm);
  taskarm->setDesiredValue(xarm);
  // Desired value for the leg (move it)
  Eigen::VectorXd xleg;
  taskleg->getSensedValue(xleg);
  xleg(0) = xleg(0) + 0.1;
  xleg(1) = xleg(1) + 0.1;
  xleg(2) = xleg(2) + 0.1;
  taskleg->setDesiredValue(xleg);

  // Solver
  double dt = 0.010;
  oscr::OSIKSolverNS solver(robot, q, dt);

  // Add tasks to the solver
  solver.pushTask(taskarm);
  solver.pushTask(taskleg);

  Eigen::VectorXd qdes;
  Eigen::VectorXd error;
  taskleg->getDerivError(error);
  error = taskleg->getError();

  // Logs
  oscr::VectorLogger arm_log("/tmp/arm.txt");
  oscr::VectorLogger leg_log("/tmp/leg.txt");
  oscr::VectorLogger legerror_log("/tmp/legerror.txt");
  oscr::VectorLogger q_log("/tmp/q.txt");
  
  unsigned int i=0;
  double time = 0.0;
  while (error.norm()>0.0005)
  {
    solver.getPositionControl(q, qdes);
    robot->updateJointConfig(q);
    q = qdes;
    error = taskleg->getError();

    // Logs
    time = i*dt;
    legerror_log.save(error, time);
    q_log.save(q, time);
    taskarm->getSensedValue(xarm);
    taskleg->getSensedValue(xleg);
    arm_log.save(xarm, time);
    leg_log.save(xleg, time);

    // Print on screen
    if (++i%50==0)
      std::cout << i << " - error norm: " << error.norm() << std::endl;
  }
  
  return 0;
}

File: examples/python/ff6dof_kine_control.py
from optparse import OptionParser
import numpy as np

from oscr.oscr import RobotModelPin, RobotModelRbdl, KineTaskPose, OSIKSolverNS
from oscr.utils import dictLink, printJointLimits, printLinkNameIDShort

if __name__ == '__main__':

    parser = OptionParser()
    parser.add_option("--back", dest="backend", default="pin",
                      help="Backend to be used: pin (default) or rbdl")
    (options, args) = parser.parse_args()
    backend = options.backend
    
    # Use a floating base?
    fbase = True
    # Robot model
    modelname = "../models/ff6dof.urdf"
    # Initialization
    if (backend=="pin"):
        print '-- Using Pinocchio as backend'
        robot = RobotModelPin(modelname, fbase)
    elif (backend=="rbdl"):
        print '-- Using RBDL as backend'
        robot = RobotModelRbdl(modelname, fbase)
    else:
        print '\nOnly pin and rbdl are accepted as backends ...',\
              "ending the program.\n"
        exit(0)

    # Link names and IDs
    linkID = dictLink(robot.mapLinkNamesIDs())
    
    # Joint configuration with all values to zero.
    q = np.zeros(robot.ndof())
    # If there is a floating base, then set a valid null quaternion
    if (robot.hasFloatingBase()):
        q[3]=1.0
    # Set the current joint configuration
    robot.updateJointConfig(q)
    # Get the current joint configuration
    qcurrent = robot.getJointConfig()

    # Definition of a task for the arm (keep) and a task for the leg
    taskarm = KineTaskPose(robot, linkID["larm3"], "position")
    taskleg = KineTaskPose(robot, linkID["lleg3"], "position")
    taskarm.setGain(1.0)
    taskleg.setGain(1.0)

    # Desired value for the arm (keep it)
    xarm = np.zeros((taskarm.getTaskDim(), 1))
    taskarm.getSensedValue(xarm)
    taskarm.setDesiredValue(xarm)
    # Desired value for the leg (move it)
    xleg = np.zeros((taskleg.getTaskDim(), 1))
    taskleg.getSensedValue(xleg)
    xleg[0] = xleg[0] + 0.1
    xleg[1] = xleg[1] + 0.1
    xleg[2] = xleg[2] + 0.1
    taskleg.setDesiredValue(xleg)
    
    # Solver
    dt = 0.010
    solver = OSIKSolverNS(robot, q, dt)

    # Add tasks to the solver
    solver.pushTask(taskarm)
    solver.pushTask(taskleg)

    qdes = np.copy(q)
    error = np.zeros((taskleg.getTaskDim(), 1))
    taskleg.getDerivError(error)
    error = taskleg.getError()

    i=0; time = 0.0
    while (np.linalg.norm(error)>0.0005):
        solver.getPositionControl(q, qdes)
        robot.updateJointConfig(q)
        q = np.copy(qdes)
        error = taskleg.getError()
        # Print on screen
        if (i%50==0):
            print i, " - error norm: ", np.linalg.norm(error)
        i = i+1
